### ISSUE TITLE
Best practice for Request Retrier with custom error serialization logic

### ISSUE THREAD

**Issue by cody1024d** [Alamofire/Alamofire#2214]
Hey guys,

I have the following scenario:

* I want to be able to parse the body of an HTTP Error (for example, a bad request response will contain a reason I can display to the user).
* I also want to, on a 401, leverage the Request Retrier to implement re-authentication

Here is the problem I'm seeing:

The Request Retrier doesn't fire unless some sort of validate is called, so that the error is set on the TaskSessionDelegate.

However, if I do want to parse the error body, the best practice is to NOT call validate, and handle the error-case within the completionHandler/custom extension.

What is the best practice then, for being able to both, call to the retrier in certain cases, AND be able to parse an HTTP error response body?

**Comment by cody1024d** [comment]
Ideally, I think, there would be a way for me to signal the Request Retrier to fire, without having to call validate() to create an error on the delegate?

**Comment by cody1024d** [comment]
As of now, I'm solving this by, in my completion handler, if the response is unsuccessful, etc., I perform the same deserialization that the responseJson method does on the body (if the body was successful).  I then have another layer up stream that converts the JsonDictionary to the expected models.

I have to say this isn't ideal, because it is a bit ugly, but I guess does work for my particular case. (I know for sure that the server will not be returning ridiculous bodies with an error)

```swift
self.manager.request(fullURL, method: httpMethod, parameters: parameters, encoding: encodingType, headers: self.headers)
                    .validate()
                    .responseJSON(completionHandler: { (response) in
                        if let value = response.result.value, let statusCode = response.response?.statusCode, response.result.isSuccess {
                            observer.onNext(NetworkResponse(statusCode: statusCode, responseBody: value))
                            observer.onCompleted()
                        } else if let error = response.result.error {
                            if let data = response.data, let statusCode = response.response?.statusCode, let jsonResponse = try? JSONSerialization.jsonObject(with: data) {
                                observer.onNext(NetworkResponse(statusCode: statusCode, responseBody: jsonResponse))
                            } else {
                                observer.onError(error)
                            }
                        } else {
                            observer.onError(NetworkError.UnexpectedResponse)
                        }
                    })
```

**Comment by cnoon** [comment]
Hi @cody1024d,

You're breakdown of what's going on here is all accurate. The request retrier **will not** be called unless the url session returned a server/connection error or validation produced an error. Currently, the request retrier is not informed of successful requests b/c...it was successful. The validation portion of a request is designed for you to inspect the response and/or the server data if necessary to determine whether or not it was actually successful.

What I'd recommend in this use case is to certainly validate at least the status code which will automatically call your retrier in the event of a 401. Then you'll want to inspect the response in the retrier to ensure it was actually an authentication error that should result in refreshing the credential.

On your point of having to deserialize the JSON twice, that is not necessary. Instead, you can build your JSON serializer to switch on the two payload types in the same parser. You certainly don't want to parse the JSON twice unless it's impossible not to.

To your last point of:

> Ideally, I think, there would be a way for me to signal the Request Retrier to fire, without having to call validate() to create an error on the delegate?

You're going to have to make a very strong case as to why this would be a good idea. This is essentially exactly what it was designed to do. Validation is a pre-hook to your response handler to generate an early error to avoid having to potentially deserialize the server payload. Saying that that entire system should not be used to generate an error so the retrier is called goes against the core principles it was designed on. It will need some good justification to be considered.

I'm going to close out this thread for now, but feel free to continue the conversation if you wish.

Cheers. 🍻

**Comment by cody1024d** [comment]
@cnoon   Thanks for the response!  You're right, that I did not end up having to serialize the Json twice, which in the very least, is a good thing :)  

--See my second response to see the exact code that I ended up going with (including a validate call).  I was able to take the raw data, and pass it through a JSONSerialization call in the case of an error.

As to my opinion on why the Retrier should fire, on second though, I was definitely incorrect.  At the time of that comment, I don't think I fully understood what was happening (and how to work around it), so my request was premature.  

*However*, I do think that there would be a use-case to allow custom validation logic, that does incorporate the error body.  This way, I would not have to make the call to the JSONSerialization method in the case of an error.  I have seen other arguments against this in other issues, though, so I do see why it wasn't done this way, as well (error bodys  are not guaranteed to be a "clean" payload, and may be gigantic).  Perhaps the option though, would still be nice.
